src/java.base/unix/classes/sun/nio/fs/UnixCopyFile.java

```diff
@@ -248,17 +279,46 @@ private static void copyFile(UnixPath source,
             // set to true when file and attributes copied
             boolean complete = false;
             try {
+                int transferSize = MIN_TRANSFER_SIZE;
+                if (transferRequiresBuffer) {
+                    int ts = MIN_TRANSFER_SIZE;
+                    try {
+                        long bss = UnixFileStoreAttributes.get(source).blockSize();
+                        long bst = UnixFileStoreAttributes.get(target).blockSize();
+                        if (bss > 0 && bst > 0) {
+                            ts = (int)(bss == bst ? bss : lcm(bss, bst));
+                        }
+                        if (ts < MIN_TRANSFER_SIZE) {
+                            int factor = (MIN_TRANSFER_SIZE + ts - 1)/ts;
+                            ts *= factor;
+                        }
+                    } catch (IllegalArgumentException | UnixException ignored) {
+                    }
+                    transferSize = ts;
```

Comment:
This code needs to move to a separate method, something like transferSize(source, target).